ObservatoryHerald/VoiceAzureCacheManager.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.IO;
+using System.Text.Json;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Xml;
+using Microsoft.CognitiveServices.Speech;
+using System.Collections.ObjectModel;
+
+namespace Observatory.Herald
+{
+    internal class VoiceAzureCacheManager
+    {
+        private string azureKey;
+        private DirectoryInfo cacheLocation;
+        private Dictionary<string, string> cacheIndex;
+        private SpeechConfig speechConfig;
+        private SpeechSynthesizer speech;
+
+
+        private string cacheIndexFile
+        {
+            get => cacheLocation.FullName + "\\VoiceIndex.json";
+        }
+
+        internal VoiceAzureCacheManager(HeraldSettings settings, HttpClient httpClient)
+        {
+            cacheLocation = new(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + "\\ObservatoryCore\\ObservatoryHerald\\");
+
+            if (!Directory.Exists(cacheLocation.FullName))
+            {
+                Directory.CreateDirectory(cacheLocation.FullName);
+            }
+
+            if (File.Exists(cacheIndexFile))
+            {
+                try
+                {
+                    cacheIndex = JsonSerializer.Deserialize<Dictionary<string, string>>(cacheIndexFile);
+                }
+                catch
+                {
+                    cacheIndex = new();
+                }
+            }
+            else
+            {
+                cacheIndex = new();
+            }
+
+            azureKey = GetAzureKey(settings, httpClient);
+            speechConfig = SpeechConfig.FromSubscription(azureKey, "eastus");
+            speech = new(speechConfig, null);
+        }
+
+        internal string GetAudioFileFromSsml(string ssml, string voice)
+        {
+            ssml = AddVoiceToSsml(ssml, voice);
+            string ssmlHash = ssml.GetHashCode().ToString("X");
+
+            string audioFile;
+
+            if (cacheIndex.ContainsKey(ssmlHash))
+            {
+                audioFile = cacheIndex[ssmlHash];
+            }
+            else
+            {
+                using var stream = RequestFromAzure(ssml);
+                audioFile = CommitToCache(ssmlHash, stream);
+            }
+
+            return audioFile;
+        }
+
+        private string CommitToCache(string ssmlHash, AudioDataStream audioData)
+        {
+            string newFile = Guid.NewGuid().ToString("D");
+            audioData.SaveToWaveFileAsync(cacheLocation + Path.DirectorySeparatorChar.ToString() + newFile + ".wav").Wait();
+            cacheIndex.Add(ssmlHash, newFile);
+            return newFile;
+        }
+
+        private AudioDataStream RequestFromAzure(string ssml)
+        {
+            var result = speech.SpeakSsmlAsync(ssml).Result;
```

Comment:
This is a remote network call, right? How are errors handled?

Reply:
An error here can't really be usefully resolved within the plugin, other than perhaps gracefully doing nothing. Was planning to update the notification handler in core to report plugin errors to a popup window, similar to journal errors during read all.

Reply:
Agree, no plugin-level resolution.

For my understanding, is that silent failure handled internally in .Result somehow? I guess I was expecting a try-catch wrapped around the async call or a conditional checking the `result` actually contains an AudioDataStream. I'd presume we wouldn't want to cache empty results.

Reply:
Currently it will throw all the way back to Core, where I'm planning to handle it in the notification event handler there to report it to a window. I'd like to avoid writing to a log file and asking users to dig for it if something goes wrong.

Reply:
Ok, SG. I assume out of scope of this PR?

Suggestion/idea: Collect in a non-modal window and perhaps in Core, provide a link/button to re-open the log window if accidentally closed.

Reply:
Yeah, out of scope here, but I hope to have it done before there's a release that include this.

Reply:
First pass on plugin error handling refactor, any thoughts? Actual messages can be refined over time as we see more of what errors can be expected.

ce3f515

Reply:
In general, LG. One question and a code simplification opportunity in comments.